File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cpp
#include <string>
#include <vector>
#include <iostream>
#include <map>
using namespace std;
map<string,int>arr;
vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    vector<int> answer;
    for(int i=0;i<terms.size();i++){
        string c=terms[i].substr(0,terms[i].find(" ")); 
        string month=terms[i].substr(terms[i].find(" ")+1, terms[i].length());
        int mon=stoi(month);
        arr[c]=mon;
    }
    for(int i=0;i<privacies.size();i++){
        string date=privacies[i].substr(0, privacies[i].find(" "));
        string category=privacies[i].substr(privacies[i].find(" ")+1, privacies[i].length());
        int year=stoi(date.substr(0,4));
        int month=stoi(date.substr(5,2));
        int day=stoi(date.substr(8));
        int end=arr[category]*28;
        int tYear=stoi(today.substr(0,4));
        int tMonth=stoi(today.substr(5,2));
        int tDay=stoi(today.substr(8));
        for(int j=0;j<end-1;j++){
            day++;
            if(day==29){
                day=1;
                month++;
            }
            if(month==13){
                month=1;
                year++;
            }
            
        }
        
        long long totalToday=28*12*tYear+tMonth*28+tDay;
        long long totalPri=28*12*year+month*28+day;
        
        if(totalToday>totalPri) answer.push_back(i+1);
        
    }
    return answer;
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;
class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        // 모든 달은 28일까지
        List<Integer> list = new ArrayList<>();
        HashMap<String, Integer> map = new HashMap<>();
        for(int i=0;i<terms.length;i++){
            String first = Character.toString(terms[i].charAt(0));
            Integer second = Integer.parseInt(terms[i].substring(2));
            map.put(first, second);
        }
        for(int i=0;i<privacies.length;i++){
            int res = map.get(Character.toString(privacies[i].charAt(privacies[i].length()-1)));
            String day = privacies[i].substring(8,10);
            int thisDay = Integer.parseInt(day);
            thisDay-=1;
            
            String dayTmp = Integer.toString(thisDay);
            String month = privacies[i].substring(5,7);
            int thisMonth = Integer.parseInt(month);
            thisMonth+=res;
            String year = privacies[i].substring(0,4);
            int thisYear = Integer.parseInt(year);
            if(thisDay==0){
                thisDay=28;
                thisMonth--;
            }
            while(true){
                if(thisMonth<=12) break;
                if(thisMonth>12){
                    thisYear++;
                    thisMonth-=12;
                }
            }
            String monthTmp = Integer.toString(thisMonth);
            String yearTmp = Integer.toString(thisYear);
            if(thisYear<Integer.parseInt(today.substring(0,4))){
                list.add(i+1);
            }
            else{
                if(thisYear==Integer.parseInt(today.substring(0,4))){
                    if(thisMonth<Integer.parseInt(today.substring(5,7))){
                        list.add(i+1);
                    }
                    else if(thisMonth==Integer.parseInt(today.substring(5,7))){
                        if(thisDay<Integer.parseInt(today.substring(8,10))){
                            list.add(i+1);
                        }
                    }
                }
            }
        }
        int[] answer = new int[list.size()];
        for(int i=0;i<list.size();i++){
            answer[i]=list.get(i);
        }
        return answer;
    }
}